components/cop_codegen/src/rpn_function.rs

```diff
@@ -62,3 +71,595 @@ impl RpnFunctionOpts {
         }
     }
 }
+
+const ARG_DEF: &str = "crate::coprocessor::dag::rpn_expr::function::ArgDef";
+const NULL: &str = "crate::coprocessor::dag::rpn_expr::function::Null";
+const RPN_FN_ARG: &str = "crate::coprocessor::dag::rpn_expr::function::RpnFnArg";
+const ARG: &str = "crate::coprocessor::dag::rpn_expr::function::Arg";
+
+pub struct RpnFnGenerator {
+    meta: Vec<Ident>,
+    item_fn: ItemFn,
+    fn_trait_ident: Ident,
+    evaluator_ident: Ident,
+    arg_types: Vec<TypePath>,
+}
+
+impl RpnFnGenerator {
+    pub fn new(meta: Vec<Ident>, item_fn: ItemFn) -> Result<Self> {
+        let arg_types = item_fn
+            .decl
+            .inputs
+            .iter()
+            .skip(meta.len()) // ctx or payload are not function args
+            .map(parse_arg_type)
+            .collect::<Result<Vec<_>>>()?;
+        let camel_name = format!("{}", item_fn.ident).to_camel_case();
+        let fn_trait_ident = parse_str(&format!("{}_Fn", camel_name)).unwrap();
+        let evaluator_ident = parse_str(&format!("{}_Evaluator", camel_name)).unwrap();
+        Ok(RpnFnGenerator {
+            meta,
+            item_fn,
+            fn_trait_ident,
+            evaluator_ident,
+            arg_types,
+        })
+    }
+
+    pub fn generate(self) -> TokenStream {
+        vec![
+            self.generate_fn_trait(),
+            self.generate_dummy_fn_trait_impl(),
+            self.generate_real_fn_trait_impl(),
+            self.generate_evaluator(),
+            self.generate_constructor(),
+            self.item_fn.into_token_stream(),
+        ]
+        .into_iter()
+        .collect()
+    }
+
+    fn generate_fn_trait(&self) -> TokenStream {
+        let (impl_generics, _, where_clause) = self.item_fn.decl.generics.split_for_impl();
+        let fn_trait_ident = &self.fn_trait_ident;
+        let (ctx_type, payload_type, result_type) = (ctx_type(), payload_type(), result_type());
+        quote! {
+            trait #fn_trait_ident #impl_generics #where_clause {
+                fn eval(
+                    self,
+                    rows: usize,
+                    ctx: &mut #ctx_type,
+                    payload: #payload_type,
+                ) -> #result_type ;
+            }
+        }
+    }
+
+    fn generate_dummy_fn_trait_impl(&self) -> TokenStream {
+        let mut generics = self.item_fn.decl.generics.clone();
+        generics
+            .params
+            .push(parse_str(&format!("D_: {}", ARG_DEF)).unwrap());
```

Comment:
I think `T: xxx` is more idiom.

Reply:
Uncommon name is used on purpose to avoid name confliction.